cmake/Utils.cmake

```diff
@@ -278,3 +278,66 @@ function(parse_cuda_version CUDA_VERSION CUDA_VERSION_MAJOR_VAR CUDA_VERSION_MIN
+# TARGET_NAME - umbrela target name used for this build
+# OUTPUT_NAME - library name used for this build
+# PREFIX - library prefix, if no provided there library will be named as ${TARGET_NAME}.python_specific_extension
+# OUTPUT_DIR - ouptut directory of the build library
+# PUBLIC_LIBS - list of libraries that should be linked in as a public one
+# PRIV_LIBS - list of libraries that should be linked in as a private one
+# SRC - list of source code files
+function(build_per_python_lib)
+    set(oneValueArgs TARGET_NAME OUTPUT_NAME OUTPUT_DIR PREFIX)
+    set(multiValueArgs PRIV_LIBS PUBLIC_LIBS SRC)
+    cmake_parse_arguments(PARSE_ARGV 0 PYTHON_LIB_ARG "${options}" "${oneValueArgs}" "${multiValueArgs}")
+
+    add_custom_target(${PYTHON_LIB_ARG_TARGET_NAME} ALL)
+
+    foreach(PYVER ${PYTHON_VERSIONS})
+
+        # check if listed python versions are accesible
+        execute_process(
+            COMMAND python${PYVER} -c "print('test', end='')"
+            OUTPUT_VARIABLE PYTHON_EXISTS)
+
+        if ("${PYTHON_EXISTS}" STREQUAL "test")
+            execute_process(
+                COMMAND python${PYVER}-config --extension-suffix
+                OUTPUT_VARIABLE PYTHON_SUFIX)
+            # remove newline and the extension
+            string(REPLACE ".so\n" "" PYTHON_SUFIX ${PYTHON_SUFIX})
+
+            execute_process(
+                COMMAND python${PYVER}-config --includes
+                OUTPUT_VARIABLE PYTHON_INCLUDES)
+            # split and make it a list
+            string(REPLACE "-I" "" PYTHON_INCLUDES ${PYTHON_INCLUDES})
+            string(REPLACE "\n" "" PYTHON_INCLUDES ${PYTHON_INCLUDES})
+            separate_arguments(PYTHON_INCLUDES)
+
+            add_library(${PYTHON_LIB_ARG_TARGET_NAME}_${PYVER} SHARED ${PYTHON_LIB_ARG_SRC})
+            target_link_libraries(${PYTHON_LIB_ARG_TARGET_NAME}_${PYVER} PUBLIC ${PYTHON_LIB_ARG_PUBLIC_LIBS})
+            target_link_libraries(${PYTHON_LIB_ARG_TARGET_NAME}_${PYVER} PRIVATE ${PYTHON_LIB_ARG_PRIV_LIBS})
+            target_link_libraries(${PYTHON_LIB_ARG_TARGET_NAME}_${PYVER} PRIVATE "-Wl,--exclude-libs,${exclude_libs}")
+
+            set_target_properties(${PYTHON_LIB_ARG_TARGET_NAME}_${PYVER}
+                                    PROPERTIES
+                                    LIBRARY_OUTPUT_DIRECTORY ${PYTHON_LIB_ARG_OUTPUT_DIR}
+                                    PREFIX "${PYTHON_LIB_ARG_PREFIX}"
+                                    OUTPUT_NAME ${PYTHON_LIB_ARG_OUTPUT_NAME}${PYTHON_SUFIX})
+            target_include_directories(${PYTHON_LIB_ARG_TARGET_NAME}_${PYVER} PRIVATE "${CMAKE_SOURCE_DIR}/third_party/pybind11/include")
+            # add includes
+            foreach(incl_dir ${PYTHON_INCLUDES})
+                target_include_directories(${PYTHON_LIB_ARG_TARGET_NAME}_${PYVER} PRIVATE ${incl_dir})
+            endforeach(incl_dir)
+
+            add_dependencies(${PYTHON_LIB_ARG_TARGET_NAME} ${PYTHON_LIB_ARG_TARGET_NAME}_${PYVER})
+        endif()
```

Comment:
How does linking with all those python version work? Don't we need to specify something?

Reply:
I don't understand the question.

Reply:
I don't see you querying for any additional python linker flags only the headers. How is that enough and how does pybind select the proper python?

Reply:
According to https://pybind11.readthedocs.io/en/stable/compiling.html#building-manually you just need the right includes.